### PR TITLE
Grant data processing

### PR DESCRIPTION
**Pull request by nonieA**
processes the labelled research fish data and EPMC abstract data to link it with the Wellcome grant data to get Wellcome grant descriptions with tool and data set labels

### CODE REVIEW

File: nutrition_labels/grant_data_processing.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import pandas as pd
+import re
+# load data
+
+epmc_tags = pd.read_csv('data/raw/EPMC_relevant_tool_pubs_manual_edit.csv')
+rf_tags = pd.read_csv('data/raw/ResearchFish/research_fish_manual_edit.csv')
+grant_tags = pd.read_csv('data/raw/wellcome-grants-awarded-2005-2019_manual_edit.csv')
+grant_data = pd.read_csv('data/raw/wellcome-grants-awarded-2005-2019.csv')
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:8]: I take it you saved the original xlsx as a csv? do you mind uploaded your csv to the [sharepoint](https://wellcomecloud.sharepoint.com/sites/DataLabs/Machine%20Learning/Forms/AllItems.aspx?newTargetListUrl=%2Fsites%2FDataLabs%2FMachine%20Learning&viewpath=%2Fsites%2FDataLabs%2FMachine%20Learning%2FForms%2FAllItems%2Easpx&viewid=a99f51bc%2D20ec%2D4b35%2D9090%2D4eedc2450aec&id=%2Fsites%2FDataLabs%2FMachine%20Learning%2FNutrition%20Labels%2Fdata%2Fraw).
Just so we are working with the same file (just in case I change the formatting by saving it as a csv in a different way to how you did)

Comment (lizgzil) [reply nutrition_labels/grant_data_processing.py:8]: can you also upload EPMC_relevant_tool_pubs_manual_edit.csv and research_fish_manual_edit.csv to the sharepoint too?

Comment (lizgzil) [reply nutrition_labels/grant_data_processing.py:8]: oo and wellcome-grants-awarded-2005-2019_manual_edit.csv

Comment (nonieA) [reply nutrition_labels/grant_data_processing.py:9]: all done :)

File: nutrition_labels/grant_data_processing.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+# sort remove list from duplicates (the later grants)
+grant_duplicates = grant_duplicates.sort_values('Award Date').drop_duplicates(subset = ['grant_number','pmid'])
+remove_rows = [i for i in remove_rows if i not in grant_duplicates.index.tolist()]
+grant_ref = grant_ref.drop(remove_rows)
+
+# finding abstract without a matching code
+no_ref = epmc_df[~epmc_df['grant_number'].isin(grant_ref['grant_number'].to_list())]
+no_ref = no_ref[~no_ref['pmid'].isin(grant_ref['pmid'].to_list())]
+missing_abstracts = epmc_tags[epmc_tags['pmid'].isin(no_ref['pmid'].to_list())] #I cant find where this grant is at all
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:92]: I think this is interesting, maybe more so the grants that weren't found though, but since you don't use it anywhere else, I'd do some print outs. Also set differences could be more clear here. e.g.
```
print("Grants flagged as relevant in EPMC publications not found in grants data: ")
print(set(epmc_df['grant_number']).difference(set(grant_ref['grant_number'])))
print("PMIDs flagged as relevant in Wellcome EPMC publications but not found to link to any grants in the grants data: ")
print(set(epmc_df['pmid']).difference(set(grant_ref['pmid'])))
```

File: nutrition_labels/grant_data_processing.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+code_list = grant_ref[['Internal ID','code','pmid']]
+
+# get rf data
+rf_useful = rf_tags[rf_tags['code '].isin([1,2,3])]
+rf_useful = rf_useful[['code ','Grant Reference','Name']]
+rf_useful = rf_useful.rename({'code ':'code','Grant Reference':'Internal ID','Name':'pmid'}, axis = 1)
+
+# merge abstract and rfs
+code_list = pd.concat([code_list,rf_useful])
+code_list.loc[:,'Internal ID'] = code_list['Internal ID'].str.strip(' ')
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:104]: `code_list.loc[:,'Internal ID']` - > `code_list['Internal ID']`

File: nutrition_labels/grant_data_processing.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+code_list_final = pd.merge(grant_data[['Internal ID','Title','Description']],code_list, how = 'inner', on = 'Internal ID')
+rf_not_in = code_list[~code_list['Internal ID'].isin(code_list_final['Internal ID'])] # six codes arent in the wellcome data
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:106]: again here I'd do a print of the results, and maybe just the grant IDs are the interesting bit
```
print("Grants flagged as relevant in RF data not found in grants data: ")
print(set(code_list['Internal ID']).difference(set(code_list_final['Internal ID'])))
```

File: nutrition_labels/grant_data_processing.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+# getting control data from wellcome grant data
+# 3== not that relevent 4==not at all relevent
+needed_cols = ['Internal ID','Title','Description','tool relevent ']
+grant_code_3 = grant_tags[needed_cols][grant_tags['tool relevent '] == 3].sample(frac = 1,random_state=4)
+grant_code_4 = grant_tags[needed_cols][grant_tags['tool relevent '] == 4].sample(frac = 1,random_state=4)
+grant_code = pd.concat([grant_code_3.iloc[:50,:],grant_code_4.iloc[:50,:]])
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:113]: let's do sampling later, since we might want to experiment with different amounts of the data/it might turn out using it all is best (which will probably be the case). so i think all this can be reduced to 
```
grant_code = grant_tags.loc[(grant_tags['tool relevent ']==4.0) | (grant_tags['tool relevent ']==3.0)]
grant_code['tool relevent ']=4
```

File: nutrition_labels/grant_data_processing.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+grant_code['tool relevent '] = 4
```
Comment (lizgzil) [inline nutrition_labels/grant_data_processing.py:114]: are 3 and 4 the irrelevant grants? If not, i think these need to be added into the training data here too

File: nutrition_labels/grant_data_processing.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+grant_ref = grant_ref.drop_duplicates(subset=['Description','grant_number','pmid'])
+
+# get remaining entrys with the same pmid grant number but different description
+grant_duplicates = grant_ref[grant_ref.duplicated(subset=['pmid','grant_number'], keep = False)]
+remove_rows = grant_duplicates.index.tolist()
+
+# sort remove list from duplicates (the later grants)
+grant_duplicates = grant_duplicates.sort_values('Award Date').drop_duplicates(subset = ['grant_number','pmid'])
+remove_rows = [i for i in remove_rows if i not in grant_duplicates.index.tolist()]
+grant_ref = grant_ref.drop(remove_rows)
```
[the remark this answers is not in the document]
Comment (nonieA) [reply nutrition_labels/grant_data_processing.py:87]: yes I thought it may be a good idea to only remove the ones for the duplicated ones but it doesnt actually matter

File: nutrition_labels/grant_data_processing.py (hunk 6 above)
Comment (nonieA) [reply nutrition_labels/grant_data_processing.py:114]: so the codes in grant_tags are 1 = super relevant, 2 = relevant, 3 = not relevant but initially looked relevant and 4 = really not relevant. They do get added to the training data later on. I'm a little bit confused what you mean here

**Comment by nonieA** [comment]
Script to process grant data for training set for naive Bayes model
